pkg/actions/nodegroup/create.go

```diff
@@ -97,100 +94,102 @@ func (m *Manager) Create(options CreateOpts, nodegroupFilter filter.NodeGroupFil
 	}
 
 	if isOwnedCluster {
-		if err := ctl.ValidateClusterForCompatibility(cfg, m.stackManager); err != nil {
+		if err := kubeProvider.ValidateClusterForCompatibility(cfg, m.stackManager); err != nil {
 			return errors.Wrap(err, "cluster compatibility check failed")
 		}
 	}
 
-	if err := vpc.ValidateLegacySubnetsForNodeGroups(cfg, ctl.Provider); err != nil {
+	if err := m.init.ValidateLegacySubnetsForNodeGroups(cfg, ctl.Provider); err != nil {
 		return err
 	}
 
-	{
-		if err := nodegroupFilter.SetOnlyLocal(ctl.Provider.EKS(), m.stackManager, cfg); err != nil {
-			return err
-		}
+	if err := m.nodeCreationTasks(options, nodegroupFilter, supportsManagedNodes, isOwnedCluster); err != nil {
+		return err
+	}
 
-		logFiltered := cmdutils.ApplyFilter(cfg, &nodegroupFilter)
-		logFiltered()
-		logMsg := func(resource string, count int) {
-			logger.Info("will create a CloudFormation stack for each of %d %s in cluster %q", count, resource, meta.Name)
-		}
-		if len(cfg.NodeGroups) > 0 {
-			logMsg("nodegroups", len(cfg.NodeGroups))
-		}
+	if err := m.postNodeCreationTasks(m.clientSet, options); err != nil {
+		return err
+	}
 
-		if len(cfg.ManagedNodeGroups) > 0 {
-			logMsg("managed nodegroups", len(cfg.ManagedNodeGroups))
-		}
+	if err := ctl.ValidateExistingNodeGroupsForCompatibility(cfg, m.stackManager); err != nil {
+		logger.Critical("failed checking nodegroups", err.Error())
+	}
 
-		if options.DryRun {
-			// Set filtered nodegroups
-			clusterConfigCopy.NodeGroups = cfg.NodeGroups
-			clusterConfigCopy.ManagedNodeGroups = cfg.ManagedNodeGroups
-			if options.ConfigFileProvided {
-				return cmdutils.PrintDryRunConfig(clusterConfigCopy, os.Stdout)
-			}
-			return cmdutils.PrintNodeGroupDryRunConfig(clusterConfigCopy, os.Stdout)
-		}
+	return nil
+}
 
-		taskTree := &tasks.TaskTree{
-			Parallel: false,
-		}
+func (m *Manager) nodeCreationTasks(options CreateOpts, nodegroupFilter filter.NodegroupFilter, supportsManagedNodes, isOwnedCluster bool) error {
+	cfg := m.cfg
+	meta := cfg.Metadata
+	init := m.init
 
-		if supportsManagedNodes && isOwnedCluster {
-			taskTree.Append(m.stackManager.NewClusterCompatTask())
-		}
+	if err := nodegroupFilter.SetOnlyLocal(m.ctl.Provider.EKS(), m.stackManager, cfg); err != nil {
+		return err
+	}
 
-		awsNodeUsesIRSA, err := eks.DoesAWSNodeUseIRSA(ctl.Provider, m.clientSet)
-		if err != nil {
-			return errors.Wrap(err, "couldn't check aws-node for annotation")
-		}
+	logFiltered := cmdutils.ApplyFilter(cfg, filter.NewNodeGroupFilter())
```

Comment:
This should be using the existing `NodeGroupFilter` that has been constructed with user-supplied options, rather than creating a new one, otherwise it'll attempt creation of nodegroups that already exist in the cluster. The newly introduced `NodegroupFilter` interface type, however, doesn't have the methods that `ApplyFilter` needs.

Reply:
Excellent feedback, thank you for finding this issue! :) Added a commit to fix this by using the user-supplied `NodegroupFilter`, then adding the missing functions to the interface. This also solved the issue of not having the full `cfg` for the nodegroup! 😄